### DISCUSSION TITLE
Why `vim.g.*` options doesn't work in a custom plugin config?

### DISCUSSION THREAD

**Discussion by amirhhashemi** [NvChad/NvChad#861] [Q&A]
Hi

`nvim-tree` has some options that I want to override and i can't do this in the setup function. like this one:
```vimscript
vim.g.nvim_tree_show_icons = {
   folders = 1,
   files = 1,
   git = 0,
}
```
I've put my custom configs in "custom.plugins.configs.nvimtree":
```lua
M.plugins = {
   default_plugin_config_replace = {
      nvim_tree = "custom.plugins.configs.nvimtree",
   },
}
```
The `vim.g.*` options don't seem to work. but when I put them in `custom/init.lua` they work. Why? 

When I pass a path to `default_plugin_config_replace` the original configs shouldn't be loaded, right?

**Comment by sillypnk** [comment]
vim.g doesn't work in init.lua for me :/

**Reply by siduck** [reply]
@Johnsiras @ahhshm thats because nvitree's config is lazy loaded and custom/init.lua is loaded before it ( see the loading order in the main init.lua file ) so nvimtree's config will load late and then that overrwrites your custom/init.lua

**Reply by sillypnk** [reply]
> @Johnsiras @ahhshm thats because nvitree's config is lazy loaded and custom/init.lua is loaded before it ( see the loading order in the main init.lua file ) so nvimtree's config will load late and then that overrwrites your custom/init.lua

it still doesn't work even tho i already did packercompile...

the nvimtree icons are that im trying to change

**Reply by sillypnk** [reply]
**chadrc.lua**
![wezterm-gui_BifZjBxggj](https://user-images.githubusercontent.com/102006493/159394954-7e5965f4-0244-495f-83e0-c2d102376c68.png)
**custom/init.lua**
![wezterm-gui_XjcczfJuzG](https://user-images.githubusercontent.com/102006493/159394996-7509a95f-c3f6-492d-927b-79c2fe49d31b.png)
![wezterm-gui_yWUhEtwZVf](https://user-images.githubusercontent.com/102006493/159395047-1271ed3b-9492-432a-81aa-bafa911bcacb.png)
Still the same

**Reply by siduck** [reply]
I think @zbirenbaum  can help. The override function just loads the table ( the default table ) i think so anything out of it like the vim globals will be ignored :/ , will try to fix

**Reply by zbirenbaum** [reply]
Edit: I just saw the parent issue, sorry I'm on mobile. That's strange, I don't use nvimtree but I'll take a look and see if there's something up here. In the meantime try the advice below.

@siduck sorry for getting back late on this, didn't notice the ping. There isn't anything to fix here. I don't have a custom/init.lua as I just use a plugin_table.lua for my custom plugins, as really that's all init.lua should be. Direct configs of plugins should never happen in there.

There are a couple of ways to fix this issue. Easiest, and most "correct", would be to copy the config function for nvimtree into a new file in custom, and add any changes you want to make in there at the end or to any existing stuff. Then just set up an override in chadrc for nvimtree pointing to that file or put a packer use statement wherever you have your custom plugins table for nvimtree that requires the file you created for the config.

Let me know if that's clear, if you still need more help I can provide some actual code but it should be pretty straightforward.

**Comment by zbirenbaum** [comment]
@ahhshm @Johnsiras @siduck I believe this is a nvim-tree issue but there is some really strange stuff going on here. It seems that the vim globals must be set before nvim-tree is required (Even the ones in the NvChad nvimtree default config appear to be having 0 effect because of this). I believe they are moving config options to lua instead of vim globals so eventually this will be fixed, but in the mean time I found a solution. 

Edit: @siduck this is not really an issue on our end but since the globals in our config are doing nothing right now, lmk if you want me to clean this up a bit and pr it, but take a look at terminals first lol :)

Place this in your init.lua or wherever your custom plugin table is:
```
{
      "kyazdani42/nvim-tree.lua",
      cmd = { "NvimTreeToggle", "NvimTreeFocus" },
      config = function()
         vim.cmd[[packadd nvim-tree.lua]]
         require("custom.plugins.overrides.nvimtree").config()
      end,
      setup = function()
         require("custom.plugins.overrides.nvimtree").setup()
         require("core.mappings").nvimtree()
      end,
},
```

Place this in custom.plugins.overrides.nvimtree (or change the paths above and put it elsewhere) and make whatever changes you like 
```
local M = {}

M.setup = function ()
   local g = vim.g
   g.nvim_tree_add_trailing = 0 -- append a trailing slash to folder names
   g.nvim_tree_git_hl = 0
   g.nvim_tree_highlight_opened_files = 0
   g.nvim_tree_indent_markers = 1
   g.nvim_tree_root_folder_modifier = table.concat { ":t:gs?$?/..", string.rep(" ", 1000), "?:gs?^??" }

   g.nvim_tree_show_icons = {
      folders = 0,
      files = 0,
      git = 0,
   }

   g.nvim_tree_icons = {
      default = "",
      symlink = "",
      git = {
         deleted = "",
         ignored = "◌",
         renamed = "➜",
         staged = "✓",
         unmerged = "",
         unstaged = "✗",
         untracked = "★",
      },
      folder = {
         default = "",
         empty = "",
         empty_open = "",
         open = "",
         symlink = "",
         symlink_open = "",
      },
   }

   g.nvim_tree_show_icons = {
      folders = 0,
      files = 0,
      git = 0,
   }
end

M.config = function ()
   local nvimtree = require("nvim-tree")
   local config = {
      filters = {
         dotfiles = false,
      },
      disable_netrw = true,
      hijack_netrw = true,
      ignore_ft_on_setup = { "dashboard" },
      auto_close = false,
      open_on_tab = false,
      hijack_cursor = true,
      hijack_unnamed_buffer_when_opening = false,
      update_cwd = true,
      update_focused_file = {
         enable = true,
         update_cwd = false,
      },
      view = {
         allow_resize = true,
         side = "left",
         width = 25,
         hide_root_folder = true,
      },
      git = {
         enable = false,
         ignore = false,
      },
      actions = {
         open_file = {
            resize_window = true,
         },
      },
   }

   nvimtree.setup(config)
end

return M
```

**Reply by siduck** [reply]
@zbirenbaum im more confused now! the override function just lets us to make changes in the default { } table of any plugin so anything outside of it such as globals wouldnt work right, Make a pr for this

**Reply by zbirenbaum** [reply]
> @zbirenbaum im more confused now! the override function just lets us to make changes in the default { } table of any plugin so anything outside of it such as globals wouldnt work right, Make a pr for this

I'm not touching the override function. I supplied a new file. One of my last PRs made it so that if you put a plugin in the custom table that is already in default the one in default is removed in favor of the one in custom. 

I'll make a PR real quick to fix the default behavior though.

**Reply by zbirenbaum** [reply]
#899

**Reply by siduck** [reply]
hmm so the pr fixes this issue? just confirming once again :thinking:

**Reply by zbirenbaum** [reply]
> hmm so the pr fixes this issue? just confirming once again :thinking: 

This fixes it in the default config. If you copy the default and change some stuff and set it as an override it works. Just test it by setting one of the show icons to 0 and notice the icons disappear.

There isn't a way to help with changing the globals outside of an override file like in chadrc or init.lua, but any override using what I wrote in the PR as a template will work fine.

**Reply by zbirenbaum** [reply]
> oh no this looks hard ._.

Why? All my PR changed was moving the vim.g globals set above where nvim-tree is required by pcall. That's all that needs to be done to make it work. 

Since setting those in the config file means values set anywhere else will be overwritten, they have to supply their own config file to override them.

**Reply by siduck** [reply]
nvm it, I didnt read your previous messages so got a lil confused
